### PR TITLE
Handle Services

### LINKED ISSUE #23: Register new works with handle provider
**ACCEPTANCE CRITERIA**

- [x] When the work is published, 
  - [x] if the work does not already have a handle, a new handle should be registered for the work
  - [x] if the work already has a handle stored in it's metadata, the handle should be verified and updated if it does not point to the correct URI for the work 
- [x] Handle registrations and updates should be performed as background jobs
- [x] Handle registrations and update errors should be logged
- [x] Any errors encountered registering or updating handles should be logged - see #24 for additional detail about how logs are accessed
- [ ] Handles should not be created until a work is published
- [x] There is a command line rake task to register handles for a work

### PR DESCRIPTION
**Pull request by no-reply**
Closes #23.

The key pieces are:

 - A `HandleBuilder` that generates handles given a `prefix` and a `hint`.

 - A `HandleRegistrar` which can `#register!` and `#update!` handles. This class is unaware of anything about `ActiveFedora::Base` objects, including whether they already have a registered handle. It knows only enough to ask about `#id`, talk to a `Handle::Connection` and use `HandleBuilder#build`.

 - A `HandleDispatcher` which knows how to assign handles to `ActiveFedora::Base#objects`. It uses an injected Registrar to register a handle and assigns it to a property on the object (default: `identifier`).

 - Two Jobs. The first to register handles, and the second to update them.

 - An Actor (`HandleAsssuranceActor`) which is appended to the actor stack and queues the delayed jobs to register a handle.

I've checked off the relevant parts of #23. I need a review to confirm that "Handles should not be created until a work is published" is satisfied. ~~The final check box (the rake task) is incoming.~~

This is ready for review, but requires a change to the configuration to point to a development handle server; so far all of this is untested against a real handle server.

### CODE REVIEW

**Comment by no-reply** [comment]
I force pushed some changes that better set up the rake task. I think this is a complete PR now, pending testing on a real handle server.

**Comment by little9** [comment]
Does the handle get indexed in Solr as a metadata attribute eventually?

**Comment by no-reply** [comment]
@little9 `HandleDispatcher` takes care of the setting (to `identifier` by default) at: https://github.com/curationexperts/epigaea/pull/125/files#diff-7a10e1b35b36c18f51c01c2763f6465aR78. The registration job and the rake task do the saving: https://github.com/curationexperts/epigaea/pull/125/files#diff-15b3794f24dd3f6e153a8307215b4327R16

File: lib/tasks/handle.rake [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        $stderr.puts "Unable to find the object for #{args[:id]}"
+        next
+      end
+
+      if object.identifier.empty?
+        handle = Tufts::HandleDispatcher.assign_for!(object: object)
+        object.save!
+        puts "Registered a handle #{handle} for #{args[:id]}"
+      else
+        puts "Identifier(s) exist for #{args[:id]}: #{object.identifiers}"
```
Comment (mark-dce) [inline lib/tasks/handle.rake]: This gives output in the form:
```
Identifier(s) exist for 7w62f8209: #<ActiveTriples::Relation:0x007f9b6a357d90>
```
I was expecting something like
```
Identifier(s) exist for 7w62f8209: ["1234.DEV/7W62F8209"]
```

File: lib/tasks/handle.rake [hunk 2]
```diff
@@ -0,0 +1,23 @@
+require 'rake'
+
+namespace :epigaea do
```
Comment (no-reply) [inline lib/tasks/handle.rake]: :ok_hand:

File: config/handle.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+development:
+  hostname: "dl.tufts.edu"
+  email: "brian.goodmon@tufts.edu"
+  prefix: 'tufts.dev'
+  admin: "0.NA/10427.TEST"
+  index: 300
+  private_key: 'path/to/key_file'
+  handle_passphrase: 'passphrase'
+
```
Comment (no-reply) [inline config/handle.yml]: :ok_hand:

File: lib/tasks/handle.rake (hunk 1 above)
Comment (no-reply) [inline lib/tasks/handle.rake]: fixed.

Comment (no-reply) [inline lib/tasks/handle.rake]: I wasn't able to test the rake task, so you might want to run it one more time before accepting the changes.

File: config/initializers/hyrax.rb [hunk 4]
```diff
@@ -24,6 +24,7 @@
 
   # Injected via `rails g hyrax:work VotingRecord`
   config.register_curation_concern :voting_record
+
```
Comment (no-reply) [inline config/initializers/hyrax.rb]: It hadn't come up before working on this, I'm inclined to just leave it here.

**Pull request merged**